File: pythonlang/Game.py
class Game:
  def __init__(self):
    self.rolls = []
    self.currentRoll = 0

  def roll(self, pins):
    self.rolls.append(pins)

  def score(self):
    score = 0
    frameIndex = 0

    for element in range(10):
      if (self.isStrike(frameIndex)):
        score += 10 + self.strikeBonus(frameIndex)
        frameIndex += 1
      elif (self.isSpare(frameIndex)):
        score += 10 + self.spareBonus(frameIndex)
        frameIndex += 2
      else:
        score += self.sumOfBallsInFrame(frameIndex)
        frameIndex += 2

    return score

  def isStrike(self, frameIndex):
    return self.rolls[frameIndex] == 10

  def sumOfBallsInFrame(self, frameIndex):
    return self.rolls[frameIndex] + self.rolls[frameIndex+1]

  def spareBonus(self, frameIndex):
    return self.rolls[frameIndex+2]

  def strikeBonus(self, frameIndex):
    return self.rolls[frameIndex+1] + self.rolls[frameIndex+2]

  def isSpare(self, frameIndex):
    return self.rolls[frameIndex] + self.rolls[frameIndex+1] == 10

File: rubylang/game.rb
class Game
  def initialize
    @rolls, @currentRoll = [], 0
  end

  def roll(pins)
    @rolls.push pins
  end

  def score
    score, frameIndex = 0, 0

    10.times do
      if (isStrike(frameIndex))
        score += 10 + strikeBonus(frameIndex)
        frameIndex += 1
      elsif (isSpare(frameIndex))
        score += 10 + spareBonus(frameIndex)
        frameIndex += 2
      else
        score += sumOfBallsInFrame(frameIndex)
        frameIndex += 2
      end
    end

    score
  end

  def isStrike(frameIndex)
    @rolls[frameIndex] == 10
  end

  def sumOfBallsInFrame(frameIndex)
    @rolls[frameIndex] + @rolls[frameIndex+1]
  end

  def spareBonus(frameIndex)
    @rolls[frameIndex+2]
  end

  def strikeBonus(frameIndex)
    @rolls[frameIndex+1] + @rolls[frameIndex+2]
  end

  def isSpare(frameIndex)
    @rolls[frameIndex] + @rolls[frameIndex+1] == 10
  end
end